### PR TITLE
Sped up shader warmup by only drawing on a 100x100 surface

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description
...and doing the draw calls on top of each other.

[big.json.gz](https://github.com/flutter/flutter/files/3408219/big.json.gz)
[little.json.gz](https://github.com/flutter/flutter/files/3408220/little.json.gz)

In big.json you'll see how the code was previous, and in little.json you'll see how it is after this PR.  This saves about 14ms in readpixels.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ x] I signed the [CLA].
- [ x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x ] I updated/added relevant documentation (doc comments with `///`).
- [x ] All existing and new tests are passing.
- [x ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/painting/shader_warm_up.dart
Comment (liyuqian) [inline packages/flutter/lib/src/painting/shader_warm_up.dart]: One reason that we added these 80.0 offsets is to visualize the warm up draws in `examples/layers/raw/shader_warm_up.dart`. Can you please add a flag here to preserve the offset in that debug app? (Maybe it's nice to do some refactor to force each draw within an 100x100 area, and automatically stack them together for actual warm-up and spread them for debugging.)

**Comment by dnfield** [comment]
Why not a 10x10 or a 1x1 surface instead?

**Review by liyuqian — COMMENTED** [review]
Great! Can you please add some tests to:
1. Verify the number of shaders being warmed up is as expected.
2. Measure the total time of shader warm-up so we can track it in our dashboard (instead of doing an observatory tracing manually every time)

**Comment by gaaclarke** [comment]
> Why not a 10x10 or a 1x1 surface instead?

The draw calls have to be onscreen.  We might be able to drop below 100x100 by tweaking the draw calls, but at 100x100 our read pixels is 4ms.  I don't know if we'll be able to drop much below that.  (Ideally we don't even read pixels but I've given up on that for now).

File: packages/flutter/lib/src/painting/shader_warm_up.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/shader_warm_up.dart]: Can you add some context here about why we think 100.0x100.0 is good?  Something like what you wrote in the response to my comment on the PR issue woul dbe good.

Comment (gaaclarke) [reply packages/flutter/lib/src/painting/shader_warm_up.dart]: Done

File: packages/flutter/lib/src/painting/shader_warm_up.dart
Comment (gaaclarke) [reply packages/flutter/lib/src/painting/shader_warm_up.dart]: Done

Comment (liyuqian) [reply packages/flutter/lib/src/painting/shader_warm_up.dart]: Can you please also set the `_drawCallSpacing` to 80.0 in `examples/layers/raw/shader_warm_up.dart`?

**Comment by liyuqian** [comment]
I realized that "adding a test to verify the number of shaders being warmed up is as expected" could be a little complicated, so I'm Ok to just have it manually  checked for this PR, and maybe add it as a test in another dedicated PR. 

On the other hand, add a test to measure the total time of shader warm-up so we can track it in our dashboard (instead of doing an observatory tracing manually every time) shouldn't be very difficult. For example, you can create `MeasureDefaultShaderWarmUp` that extends `DefaultShaderWarmUp` and overrides the `execute` like this:
```
class MeasureDefaultShaderWarmUp extends DefaultShaderWarmUp {
  Duration measuredTime;

  @override
  Future<void> execute() async {
    final DateTime start = DateTime.now();
    await super.execute();
    measuredTime = DateTime.now() - start;
  }
}
```
You can then write a device lab driver test to report the measurement: https://github.com/flutter/flutter/tree/master/dev/devicelab

Please let me know if you have any questions about the device lab test. Ideally, maybe we can land the driver test before the merge of this PR so we can get an official number of how much time is saved on our test devices :smile:

**Comment by gaaclarke** [comment]
> You can then write a device lab driver test to report the measurement: https://github.com/flutter/flutter/tree/master/dev/devicelab

Hey Yuqian, I looked into this and it doesn't seem trivial.  Whatever tests covered this code previously should cover this PR.  I'm not sure if continual tracking of shader warmup is something we want to do, considering the effort to do so.  This is demonstrably faster with profiling.  I agree that the benchmark would be nice to have, but I'm failing to see how it is a good return on investment.  Let's chat tomorrow.

File: packages/flutter/lib/src/painting/shader_warm_up.dart
Comment (liyuqian) [reply packages/flutter/lib/src/painting/shader_warm_up.dart]: @gaaclarke : any comment on this?

Comment (gaaclarke) [reply packages/flutter/lib/src/painting/shader_warm_up.dart]: Done.

**Review by liyuqian — APPROVED** [review]
LGTM modulo the analyze warnings.

**Comment by gaaclarke** [comment]
The 2 outstanding tests have passed, they just aren't reporting so here for some reason.

**Pull request merged**
